sycl/unittests/helpers/PiMock.hpp

```diff
@@ -0,0 +1,208 @@
+//    from underlying runtimes;
+// 3) testing the order of PI API calls;
+// ..., etc.
+//
+//===----------------------------------------------------------------------===//
+
+#include <CL/sycl.hpp>
+#include <CL/sycl/detail/common.hpp>
+#include <CL/sycl/detail/pi.hpp>
+#include <detail/platform_impl.hpp>
+
+#include <functional>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace unittest {
+
+namespace detail = cl::sycl::detail;
+namespace RT = detail::pi;
+
+/// Overwrites the input PiPlugin's PiFunctionTable entry for the given PI API
+/// with a given function pointer.
+///
+/// \param MPlugin is a pointer to the PiPlugin instance that will be modified
+/// \param FuncPtr is a pointer to the function that will override the original
+///        function table entry
+#define _PI_API(api)                                                           \
+  template <detail::PiApiKind PiApiOffset>                                     \
+  void setFuncPtr(RT::PiPlugin *MPlugin, decltype(&::api) FuncPtr);            \
+  template <>                                                                  \
+  void setFuncPtr<detail::PiApiKind::api>(RT::PiPlugin * MPlugin,              \
+                                          decltype(&::api) FuncPtr) {          \
+    MPlugin->PiFunctionTable.api = FuncPtr;                                    \
+  }
+#include <CL/sycl/detail/pi.def>
+
+/// The PiMock class wraps an instance of a SYCL platform class,
+/// and manages all mock redefinitions for the underlying plugin.
+///
+/// Mock platform instances must not share the plugin resources with
+/// any other SYCL platform within the given context. Otherwise, mock
+/// redefinitions would also affect other platforms' behavior.
+/// Therefore, any plugin-related information is fully copied whenever
+/// a user-passed SYCL object instance is being mocked.
+/// The underlying SYCL platform must be a non-host plaftorm to facilitate
+/// plugin usage.
+///
+/// Simple usage examples would look like this:
+/// ```
+/// pi_result redefinePiProgramRetain(pi_program program) { /*code*/ }
+/// /*...*/
+/// unittest::PiMock Mock;
+/// Mock.redefine<PiApiKind::piProgramRetain>(redefinePiProgramRetain);
+/// platform &MockP = Mock.getPlatform();
+/// /*...*/
+/// ```
+/// ```
+/// pi_result redefinePiProgramRetain(pi_program program) { /*code*/ }
+/// /*...*/
+/// queue Q;
+/// unittest::PiMock Mock(Q);
+/// Mock.redefine<PiApiKind::piProgramRetain>(redefinePiProgramRetain);
+/// Q.submit(/* expecting mock behavior */);
+/// /*...*/
+/// ```
+class PiMock {
+public:
+  /// Default-constructs PiMock.
+  ///
+  /// Delegates to the device_selector-based constructor by passing
+  /// a default_constructor and assuming a non-host device will be
+  /// available.
+  PiMock() : PiMock(cl::sycl::default_selector{}) {}
+
+  /// Constructs PiMock from a queue.
+  ///
+  /// \param Queue is a reference to a SYCL queue to which
+  ///        the mock redefinitions will apply
```

Comment:
```suggestion
  ///        the mock redefinitions will apply.
```

Reply:
I think this is how LLVM coding standards recommends doing it https://llvm.org/docs/CodingStandards.html#doxygen-use-in-documentation-comments. But that’s definitely a minor comment.